README.md
# Universally Unique Lexicographically Sortable Identifier

This is a re-implementation of the ULID library by [Airframe](https://github.com/wvlet/airframe) in Kotlin. It follows
the [ULID spec](https://github.com/ulid/spec).

## Advantages

- 128-bit compatibility with UUID
- 1.21e+24 unique ULIDs per millisecond
- Lexicographically sortable!
- Canonically encoded as a 26 character string, as opposed to the 36 character UUID
- Uses Crockford's base32 for better efficiency and readability (5 bits per character)
- Case-insensitive
- No special characters (URL safe)
- Monotonic sort order (correctly detects and handles the same millisecond)

## API

```kotlin
// Constructors returning `ULID` objects
val newRandomUlid: ULID = ULID.newULID()
val newRandomUlidStr: String = ULID.newULIDString()

// Constructors that can fail and return Either<Error, ULID>
val ulidFromString = ULID("01BX5ZZKBKACTAV9WEVGEMMVRZ")
val ulidFromUUID = ULID.fromUUID(UUID.randomUUID())
val ulidFromUUIDString = ULID.fromUUIDString("f93f9fa5-f760-4341-b62b-508def86f087")
val ulidFromBytes = ULID.fromBytes(byteArray)
val ulidFromRawData = ULID.of(unixTimeMillis, randHi, randLow)

// Usage
val ulid = ULID.newULID()
println(ulid)                   // 01F51KNV65BQH43X6FMZ3JJT66
println(ulid.epochMillis)       // 1620330605765
println(ulid.toInstant())       // 2021-05-06T19:50:05.765Z
println(ulid.toUUID())          // 0179433a-ecc5-5de2-41f4-cfa7c72968c6
println(ulid.toBytes().asHex()) // 01 79 43 3a ec c5 5d e2 41 f4 cf a7 c7 29 68 c6
println(ULID.isValid("deadbeef"))                   // false
println(ULID.isValid("01F51KNV65BQH43X6FMZ3JJT66")) // true
```

### Coroutine API

ULID generation needs to be synchronized to allow correct handling of sub-millisecond generation. The initial airframe
implementation uses `synchronized(this)` on a static singleton, and the same pattern was followed here. To accommodate
the unlikely event of this blocking for a long time, a suspending API without the use of `synchronized` is available:

```kotlin
import com.kobil.ulid.suspending.newULID

suspend fun businessLogic() {
  val ulid = ULID.newULID() // 'suspendable' construction of ULID here
}
```

Creating ULIDs is synchronized using a Kotlin `Mutex`, which offers more fine-grained synchronization control at a higher
cost, so using the `suspend` API will degrade throughput to about the same level as `java.util.UUID`.

## Performance

Extensive performance review has not yet been performed, but a basic single-threaded throughput test comparing different
existing ULID libraries is discussed here.

The tests were performed on an 2.8 GHz Core i7-7700HQ (MacBook Pro 14,3 2017), each generating 1 million IDs after a
warm-up of 1000 IDs.

```
com.github.guepardoapps.kulid time: 1330 ms
          wvlet.airframe.ulid time: 1040 ms
        de.huxhorn.sulky.ulid time: 9856 ms
               java.util.UUID time: 676  ms
               com.kobil.ulid time: 545  ms
```

## Building

Tests:

```
./gradlew clean test
```

Packaging

```
./gradlew clean assemble
```

Publishing (don't forget the appropriate credentials in `gradle.properties`)

```
./gradlew publishAllPublicationsToNexusRepository
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.2.2] - 2022-09-06

### Changed

- Internal build process changes

## [1.2.1] - 2022-06-18

### Added

- Added `nullValue` constant to the `ULID` class for easier comparison in library users.

##  [1.2.0] - 2022-06-18

### Changed

- Constructors that can fail (i.e., accepting external input) now return `Either<Error, ULID>`

## [1.1.1] - 2022-05-28

### Fixed

- Use `SecureRandom()` instead of `SecureRandom.getInstanceStrong()` to avoid blocking when entropy is low

## [1.1.0] - 2022-05-28

### Added

- Suspending API for `newULID` and `newULIDString`

## [1.0.3] - 2022-05-21

### Added

- fromUUIDString/toUUIDString functions

## [1.0.2] - 2022-05-14

### Added

- Equals/hashcode implementation to fix equality bug

## [1.0.1] - 2022-05-06

### Fixed

- Use SecureRandom#nextBytes instead of #generateSeed

## [1.0.0] - 2022-05-06

### Added

- Initial release


src/test/kotlin/com/kobil/ulid/ULIDTest.kt
package com.kobil.ulid

import arrow.core.Either
import com.kobil.ulid.blocking.newULID
import io.kotest.assertions.arrow.core.shouldBeRight
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.FunSpec
import io.kotest.data.forAll
import io.kotest.matchers.comparables.shouldBeEqualComparingTo
import io.kotest.matchers.comparables.shouldBeLessThanOrEqualTo
import io.kotest.matchers.comparables.shouldNotBeEqualComparingTo
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import java.util.UUID
import com.kobil.ulid.suspending.newULID as suspendingNewULID

class ULIDTest : FunSpec({
  test("Check basic construction") {
    for (i in 1..10) {
      val ulid = ULID.newULID()
      val str = ulid.toString()

      val parsed = ULID.fromString(str).shouldBeRight()
      parsed shouldBeEqualComparingTo ulid
      parsed shouldBeLessThanOrEqualTo ULID.maxValue
    }
  }

  test("Check toString/toBytes") {
    forAll<Long, Long, Long> { a, rh, rl ->
      val unixTime = a and ((0L).inv().shr((64 - 48)))
      // Identity
      val ulid = ULID.of(unixTime, rh, rl).shouldBeRight()

      ulid.compareTo(ulid) shouldBe 0
      ULID.fromString(ulid.toString()) shouldBe ulid
      ULID.fromBytes(ulid.toBytes()) shouldBe ulid

      // Basic conditions
      ulid.epochMillis shouldBe unixTime
      ulid.timestamp shouldBe unixTime
      ulid.randomness() shouldBe Pair(rh and 0xffffL, rl)
      ULID.isValid(ulid.toString()) shouldBe true
    }
  }

  test("Check generating monotonically increasing ULIDs") {
    val start = System.currentTimeMillis() - 1
    val lst = (0..10000).map { ULID.newULID() }
    val end = System.currentTimeMillis()
    lst.forEach { x ->
      (x.epochMillis in start..end) shouldBe true
    }

    lst.windowed(2).all { pair ->
      pair[0] < pair[1] &&
        pair[0].epochMillis <= pair[1].epochMillis
    } shouldBe true
  }

  test("Check epochMillis should equal unixTime") {
    forAll<Long> { timeMillis ->
      val unixTime = timeMillis and 0xffffffffffffL
      ULID.of(unixTime, 0, 0).shouldBeRight().epochMillis shouldBe unixTime
    }
  }

  test("Check constructors") {

    ULID.of(ULID.minTime, 0, 0).shouldBeRight() shouldBeEqualComparingTo ULID("00000000000000000000000000")
    ULID.of(ULID.minTime, 0, 0).shouldBeRight() shouldBeEqualComparingTo ULID.nullValue
    ULID.of(1L, 0, 0).shouldBeRight() shouldBeEqualComparingTo ULID("00000000010000000000000000")
    ULID.of(ULID.maxTime, 0, 0).shouldBeRight() shouldBeEqualComparingTo ULID("7ZZZZZZZZZ0000000000000000")
    ULID.of(ULID.maxTime, 0L.inv(), 0L.inv()).shouldBeRight() shouldBeEqualComparingTo ULID.maxValue
    ULID.of(0L, 0, 0L.inv()).shouldBeRight() shouldBeEqualComparingTo ULID("0000000000000FZZZZZZZZZZZZ")
    ULID.of(0L, 0L.inv(), 0L.inv()).shouldBeRight() shouldBeEqualComparingTo ULID("0000000000ZZZZZZZZZZZZZZZZ")
  }

  test("Check Validation") {
    shouldThrow<IllegalArgumentException> {
      // testing the private constructor
      ULID("deafbeef")
    }
    val u1 = ULID.of(ULID.minTime - 1L, 0, 0)
    val u2 = ULID.of(ULID.maxTime + 1L, 0, 0)

    (u1 as Either.Left).value shouldBe Error.TimeOutOfBounds
    (u2 as Either.Left).value shouldBe Error.TimeOutOfBounds
  }

  test("Check encoded timestamp should be correct") {
    val ulid = ULID.newULID()
    val ts = ulid.epochMillis
    val tsString = ulid.toString().substring(0, 10)
    val decodedTs = CrockfordBase32.decode48bits(tsString)
    ts shouldBe decodedTs
  }

  test("Check from UUID") {
    val uuid = UUID.randomUUID()
    val ulid = ULID.fromUUID(uuid).shouldBeRight()
    ulid.toUUID() shouldBeEqualComparingTo uuid
  }

  test("Check from UUID String") {
    val uuidString = UUID.randomUUID().toString()
    val ulid = ULID.fromUUIDString(uuidString).shouldBeRight()
    ulid.toUUID().toString() shouldBeEqualComparingTo uuidString
  }

  test("Check to UUID") {
    val ulid = ULID.newULID()
    val uuid = ulid.toUUID()
    ULID.fromUUID(uuid).shouldBeRight() shouldBeEqualComparingTo ulid
  }

  test("Check to UUID String") {
    val ulid = ULID.newULID()
    val uuidString = ulid.toUUIDString()
    uuidString shouldBeEqualComparingTo ulid.toUUID().toString()
  }

  test("Check equality and comparisons") {
    val ulid1 = ULID.fromString("01F5PCMVASGS6PWN00F9VX3TEK") as Either.Right
    val ulid2 = ULID.fromString("01F5PCMVASGS6PWN00F9VX3TEK") as Either.Right
    val ulid3 = ULID.fromString("01F5PCMVATGS6PWN00F9VX3TEK") as Either.Right // timestamp + 1

    ulid1.value shouldBeEqualComparingTo ulid2.value
    ulid1 shouldBe ulid2
    ulid2 shouldNotBe ulid3
    ulid2.value shouldNotBeEqualComparingTo ulid3.value
    (ulid1.value < ulid2.value) shouldBe false
    (ulid1.value > ulid2.value) shouldBe false
    (ulid2.value < ulid3.value) shouldBe true
    (ulid2.value > ulid3.value) shouldBe false
  }

  test("Check to/from bytes") {
    val ulid = ULID.newULID()
    val bytes = ulid.toBytes()
    ULID.fromBytes(bytes).shouldBeRight() shouldBeEqualComparingTo ulid
  }

  test("Check parallel suspend constructor") {
    val ulids = (1..1000).map {
      // Be aware that this is aliased in this file to "suspendingNewULID" to avoid name clashes with the blocking
      // version used in all other tests
      ULID.suspendingNewULID()
    }
    ulids.zipWithNext().forEach {
      (it.first < it.second) shouldBe true
    }
  }
})


src/test/kotlin/com/kobil/ulid/CrockfordBase32Test.kt
package com.kobil.ulid

import io.kotest.core.spec.style.FunSpec
import io.kotest.data.forAll
import io.kotest.matchers.shouldBe

class CrockfordBase32Test : FunSpec({
  test("Encoding Long Pairs") {
    forAll<Long, Long> { hi, low ->
      val encoded = CrockfordBase32.encode128bits(hi, low)
      val (hi_d, low_d) = CrockfordBase32.decode128bits(encoded)
      Pair(hi, low) shouldBe Pair(hi_d, low_d)
    }
  }
})


src/main/kotlin/com/kobil/ulid/ULID.kt
package com.kobil.ulid
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*
 * This is forked from https://github.com/wvlet/airframe/blob/master/airframe-ulid/
 */

import arrow.core.Either
import arrow.core.left
import arrow.core.right
import kotlinx.coroutines.delay
import java.nio.ByteBuffer
import java.security.SecureRandom
import java.time.Instant
import java.util.UUID
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicReference

class ULID internal constructor(private val ulid: String) : Comparable<ULID> {

  init {
    // This is a basic sanity check. The constructor is `internal` and therefore not part of the API
    require(ulid.length == 26) { "Private constructor called with sting of invalid length" }
  }

  override fun toString(): String {
    return ulid
  }

  val epochMillis: Long by lazy { CrockfordBase32.decode48bits(ulid.substring(0, 10)) }

  val timestamp = epochMillis

  fun randomness(): Pair<Long, Long> {
    val (hi, low) = CrockfordBase32.decode128bits(ulid)
    return Pair(hi and 0xffffL, low)
  }

  fun toInstant(): Instant {
    return Instant.ofEpochMilli(epochMillis)
  }

  fun toUUID(): UUID {
    val (hi, low) = CrockfordBase32.decode128bits(ulid)
    return UUID(hi, low)
  }

  fun toUUIDString(): String {
    return toUUID().toString()
  }

  fun toBytes(): ByteArray {
    val (hi, low) = CrockfordBase32.decode128bits(ulid)
    val b = ByteArray(16)
    for (i in 0..8) {
      b[i] = ((hi.ushr(64 - (i + 1) * 8)) and 0xffL).toByte()
    }
    for (i in 0..7) {
      b[i + 8] = ((low.ushr(64 - (i + 1) * 8)) and 0xffL).toByte()
    }
    return b
  }

  override fun compareTo(other: ULID): Int {
    return this.ulid.compareTo(other.ulid)
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (javaClass != other?.javaClass) return false

    other as ULID

    if (ulid != other.ulid) return false
    if (timestamp != other.timestamp) return false

    return true
  }

  override fun hashCode(): Int {
    var result = ulid.hashCode()
    result = 31 * result + timestamp.hashCode()
    return result
  }

  companion object {
    const val minTime = 0L
    const val maxTime = (0L.inv()).ushr(64 - 48) // Timestamp uses 48-bit range
    val maxValue: ULID = ULID("7ZZZZZZZZZZZZZZZZZZZZZZZZZ")
    val nullValue: ULID = ULID("00000000000000000000000000")

    private val random = SecureRandom()

    internal val defaultGenerator = ULIDGenerator { bs -> random.nextBytes(bs); bs }

    fun fromString(ulid: String): Either<Error, ULID> {
      if (ulid.length != 26) {
        return Error.IncorrectStringLength.left()
      }
      if (CrockfordBase32.isValidBase32(ulid).not()) {
        return Error.NonBase32Char.left()
      }
      return ULID(ulid).right()
    }

    fun fromUUID(uuid: UUID): Either<Error, ULID> {
      val b = ByteBuffer.wrap(ByteArray(16))
      b.putLong(uuid.mostSignificantBits)
      b.putLong(uuid.leastSignificantBits)
      return fromBytes(b.array())
    }

    fun fromUUIDString(uuid: String): Either<Error, ULID> {
      return fromUUID(UUID.fromString(uuid))
    }

    /**
     * Create an ULID from a given timestamp (48-bit) and a random value (80-bit)
     * @param unixTimeMillis 48-bit unix time millis
     * @param randHi  16-bit hi-part of 80-bit random value
     * @param randLow 64-bit low-part of 80-bit random value
     * @return
     */
    fun of(unixTimeMillis: Long, randHi: Long, randLow: Long): Either<Error, ULID> {
      if (unixTimeMillis < 0L || unixTimeMillis > maxTime) {
        return Error.TimeOutOfBounds.left()
      }
      val hi: Long = (unixTimeMillis.shl(64 - 48)) or (randHi and 0xffff)
      val low: Long = randLow
      return ULID(CrockfordBase32.encode128bits(hi, low)).right()
    }

    /**
     * Create a ne ULID from a byte sequence (16-bytes)
     */
    fun fromBytes(bytes: ByteArray): Either<Error, ULID> = fromBytes(bytes, 0)

    /**
     * Create a ne ULID from a byte sequence (16-bytes)
     */
    fun fromBytes(bytes: ByteArray, offset: Int): Either<Error, ULID> {
      if (offset + 16 > bytes.size) {
        return Error.BytesOutOfBounds.left()
      }

      var i = 0
      var hi = 0L
      while (i < 8) {
        hi = hi.shl(8)
        hi = hi or (bytes[offset + i].toLong() and 0xffL)
        i += 1
      }
      var low = 0L
      while (i < 16) {
        low = low.shl(8)
        low = low or (bytes[offset + i].toLong() and 0xffL)
        i += 1
      }
      return ULID(CrockfordBase32.encode128bits(hi, low)).right()
    }

    /**
     * check a given string is valid as ULID
     * @param ulid
     * @return
     */
    fun isValid(ulid: String): Boolean {
      return ulid.length == 26 && CrockfordBase32.isValidBase32(ulid)
    }

    /**
     * ULID generator.
     * @param rng a function that returns a 80-bit random values in ByteArray (size:10)
     */
    internal class ULIDGenerator(val rng: (ByteArray) -> ByteArray) {
      private val baseSystemTimeMillis = System.currentTimeMillis()
      private val baseNanoTime = System.nanoTime()

      private val lastValue = AtomicReference(Pair(0L, 0L))

      private fun currentTimeInMillis(): Long {
        // Avoid unexpected rollback of the system clock
        return baseSystemTimeMillis + TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - baseNanoTime)
      }

      /**
       * Generate ULID string. This method is synchronized so that only a single-thread can generate ULID based on the
       * previous value
       *
       * Tips for optimizing performance:
       *
       * 1. Reduce the number of Random number generation. SecureRandom is quite slow, so within the same milliseconds, just incrementing the randomness part will provide
       * better performance.
       * 2. Generate random in Array[Byte] (10 bytes = 80 bits). Regular Random uses 48-bit seed, so calling Random.nextInt (32 bits) x 3 is faster, but
       * SecureRandom has optimization for Array[Byte] generation, which is much faster than calling nextInt three times.
       * 3. ULIDs are often used in the string value form (e.g., transaction IDs, object IDs which can be embedded to URLs, etc.). Generating ULID String from the beginning
       * is ideal.
       * 4. In base32 encoding/decoding, use bit-shift operators as much as possible to utilize CPU registers and memory cache.
       */
      fun generateBlocking(): String {
        val unixTimeMillis: Long = currentTimeInMillis()
        if (unixTimeMillis > maxTime) {
          throw IllegalStateException("The current time is outside range (48bit). This error cannot be recovered from.")
        }

        // Add a guard so that only a single-thread can generate ULID based on the previous value
        synchronized(this) {
          val (hi, low) = lastValue.get()
          val lastUnixTime = (hi.ushr(16)) and 0xffffffffffffL
          if (lastUnixTime == unixTimeMillis) {
            // do increment
            return if (low != 0L.inv()) {
              generateFrom(hi, low + 1L)
            } else {
              var nextHi = (hi and (0L.inv().shl(16)).inv()) + 1
              if ((nextHi and (0L.inv().shl(16))) != 0L) {
                // Random number overflow. Wait for one millisecond and retry
                Thread.sleep(1)
                generateBlocking()
              } else {
                nextHi = nextHi or unixTimeMillis.shl(64 - 48)
                generateFrom(nextHi, 0)
              }
            }
          } else {
            // No conflict at millisecond level. We can generate a new ULID safely
            val bs = ByteArray(10)
            return generateFrom(unixTimeMillis, rng(bs))
          }
        }
      }

      /**
       * Generate ULID String in a coroutine. This call needs to be guarded by a mutex, because ULID generation needs
       * to happen sequentially to assure ordering
       */
      internal suspend fun generate(): String {
        val unixTimeMillis: Long = currentTimeInMillis()
        if (unixTimeMillis > maxTime) {
          throw IllegalStateException("The current time is outside range (48bit). This error cannot be recovered from.")
        }

        val (hi, low) = lastValue.get()
        val lastUnixTime = (hi.ushr(16)) and 0xffffffffffffL
        if (lastUnixTime == unixTimeMillis) {
          // do increment
          return if (low != 0L.inv()) {
            generateFrom(hi, low + 1L)
          } else {
            var nextHi = (hi and (0L.inv().shl(16)).inv()) + 1
            if ((nextHi and (0L.inv().shl(16))) != 0L) {
              // Random number overflow. Wait for one millisecond and retry
              delay(1)
              generate()
            } else {
              nextHi = nextHi or unixTimeMillis.shl(64 - 48)
              generateFrom(nextHi, 0)
            }
          }
        } else {
          // No conflict at millisecond level. We can generate a new ULID safely
          val bs = ByteArray(10)
          return generateFrom(unixTimeMillis, rng(bs))
        }
      }

      private fun generateFrom(unixTimeMillis: Long, rand: ByteArray): String {
        // We need a 80-bit random value here.
        if (rand.size != 10) {
          throw IllegalStateException(
            "generateFrom was called with an incorrect size of random bytes, meaning the maintainer messed something up"
          )
        }

        val hi = ((unixTimeMillis and 0xffffffffffffL).shl(64 - 48)) or
          (rand[0].toLong() and 0xffL).shl(8) or (rand[1].toLong() and 0xffL)
        val low: Long =
          ((rand[2].toLong() and 0xffL).shl(56)) or
            ((rand[3].toLong() and 0xffL).shl(48)) or
            ((rand[4].toLong() and 0xffL).shl(40)) or
            ((rand[5].toLong() and 0xffL).shl(32)) or
            ((rand[6].toLong() and 0xffL).shl(24)) or
            ((rand[7].toLong() and 0xffL).shl(16)) or
            ((rand[8].toLong() and 0xffL).shl(8)) or
            (rand[9].toLong() and 0xffL)
        return generateFrom(hi, low)
      }

      private fun generateFrom(hi: Long, low: Long): String {
        lastValue.set(Pair(hi, low))
        return CrockfordBase32.encode128bits(hi, low)
      }
    }
  }
}


src/main/kotlin/com/kobil/ulid/CrockfordBase32.kt
package com.kobil.ulid

/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*
 * This is forked from https://github.com/wvlet/airframe/blob/master/airframe-ulid/
 */

object CrockfordBase32 {
  private val ENCODING_CHARS = arrayOf(
    '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'J', 'K', 'M', 'N', 'P',
    'Q', 'R', 'S', 'T', 'V', 'W', 'X', 'Y', 'Z'
  )

  private val DECODING_CHARS = arrayOf<Byte>(
    -1, -1, -1, -1, -1, -1, -1, -1, // 0
    -1, -1, -1, -1, -1, -1, -1, -1, // 8
    -1, -1, -1, -1, -1, -1, -1, -1, // 16
    -1, -1, -1, -1, -1, -1, -1, -1, // 24
    -1, -1, -1, -1, -1, -1, -1, -1, // 32
    -1, -1, -1, -1, -1, -1, -1, -1, // 40
    0, 1, 2, 3, 4, 5, 6, 7, // 48
    8, 9, -1, -1, -1, -1, -1, -1, // 56
    -1, 10, 11, 12, 13, 14, 15, 16, // 64
    17, 1, 18, 19, 1, 20, 21, 0, // 72
    22, 23, 24, 25, 26, -1, 27, 28, // 80
    29, 30, 31, -1, -1, -1, -1, -1, // 88
    -1, 10, 11, 12, 13, 14, 15, 16, // 96
    17, 1, 18, 19, 1, 20, 21, 0, // 104
    22, 23, 24, 25, 26, -1, 27, 28, // 112
    29, 30, 31 // 120
  )

  fun decode(ch: Char): Byte {
    return DECODING_CHARS[(ch.code and 0x7f)]
  }

  fun encode(i: Int): Char {
    return ENCODING_CHARS[i and 0x1f]
  }

  fun decode128bits(s: String): Pair<Long, Long> {
    /**
     * |      hi (64-bits)     |    low (64-bits)    |
     */
    val len = s.length
    if (len != 26) {
      throw IllegalArgumentException("String length must be 26: $s (length: $len)")
    }
    var i = 0
    var hi = 0L
    var low = 0L

    val carryMask = ((0L).inv().ushr(5)).inv()
    while (i < 26) {
      val v = decode(s[i])
      val carry = (low and carryMask).ushr(64 - 5)
      low = low.shl(5)
      low = low or v.toLong()
      hi = hi.shl(5)
      hi = hi or carry
      i += 1
    }
    return Pair(hi, low)
  }

  fun encode128bits(hi: Long, low: Long): String {
    val s = StringBuilder(26)
    var i = 0
    var h = hi
    var l = low
    // encode from lower 5-bit
    while (i < 26) {
      s.append(encode((l and 0x1fL).toInt()))
      val carry = (h and 0x1fL).shl(64 - 5)
      l = l.ushr(5)
      l = l or carry
      h = h.ushr(5)
      i += 1
    }
    return s.reverse().toString()
  }

  fun decode48bits(s: String): Long {
    val len = s.length
    if (len != 10) {
      throw IllegalArgumentException("String size must be 10: $s (length:$len)")
    }
    var l: Long = decode(s[0]).toLong()
    var i = 1
    while (i < len) {
      l = l.shl(5)
      l = l or decode(s[i]).toLong()
      i += 1
    }
    return l
  }

  fun isValidBase32(s: String): Boolean {
    return s.all { decode(it) != (-1).toByte() }
  }
}
